### ISSUE TITLE
Proposal of converting MySQL @variables to DECLARE block of PostgreSQL

### ISSUE THREAD

**Issue by krasandr** [darold/ora2pg#590]
My idea of proposal for converting MySQL @variables to PostgreSQL is (for example, MySQL function before converting):
```
| F_ConvertHormone | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` FUNCTION `F_ConvertHormone`(HormoneName VARCHAR(255), HormoneUnit VARCHAR(255), 
  HormoneValue double(17, 4)) RETURNS varchar(255) CHARSET utf8
BEGIN

  SET @ConvertOperation := 0;
  SET @ConvertFactor := 0;
  SELECT
    f.HormoneOperation,
    f.HormoneFactor INTO @ConvertOperation, @ConvertFactor
  FROM ivf_lab_formulas f
  WHERE f.HormoneName = HormoneName
  AND f.HormoneUnit = HormoneUnit;

  IF @ConvertOperation = 0 THEN 
    RETURN HormoneValue;
  END IF;

  IF @ConvertOperation = 1 THEN 
    RETURN HormoneValue + @ConvertFactor;
  END IF;

  IF @ConvertOperation = 2 THEN 
    RETURN HormoneValue - @ConvertFactor;
  END IF;

  IF @ConvertOperation = 3 THEN 
    RETURN HormoneValue * @ConvertFactor;
  END IF;

  IF @ConvertOperation = 4 THEN 
    RETURN HormoneValue / @ConvertFactor;
  END IF;

  RETURN 0;
END | utf8                 | utf8_general_ci      | utf8_general_ci    |
```
Right now, after conversion, Ora2Pg produces this output:
```
CREATE OR REPLACE FUNCTION f_converthormone (HormoneName varchar(255), HormoneUnit varchar(255), HormoneValue double precision(17, 4)) RETURNS varchar AS $body$
BEGIN

  SET @ConvertOperation := 0;
  SET @ConvertFactor := 0;
  SELECT
    f.HormoneOperation,
    f.HormoneFactor INTO @ConvertOperation, @ConvertFactor
  FROM ivf_lab_formulas f
  WHERE f.HormoneName = HormoneName
  AND f.HormoneUnit = HormoneUnit;

  IF @ConvertOperation = 0 THEN 
    RETURN HormoneValue;
  END IF;

  IF @ConvertOperation = 1 THEN 
    RETURN HormoneValue + @ConvertFactor;
  END IF;

  IF @ConvertOperation = 2 THEN 
    RETURN HormoneValue - @ConvertFactor;
  END IF;

  IF @ConvertOperation = 3 THEN 
    RETURN HormoneValue * @ConvertFactor;
  END IF;

  IF @ConvertOperation = 4 THEN 
    RETURN HormoneValue / @ConvertFactor;
  END IF;

  RETURN 0;
END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION f_converthormone (HormoneName varchar(255), HormoneUnit varchar(255), HormoneValue double precision(17, 4)) FROM PUBLIC;
```
There is 4 possibilities for MySQL @variables:
1) it is variable, that is declared in parameter list of function/procedure (or returns part), so you know it's type and you can set it right, if they are used with @variable syntax
2) it is not declared, but variable sets it's type, based on basic assignment, like @a := 0 or @b := "x"
(maybe you can guess it't type, based on basic checks, but it is no so mandatory in first fix)
3) it is not declared, but variable sets it's type, based on some query (so, because you don't keep right context of every column of each table, or query is complex and must be evaluated with some grammar),
so here you can invent some keyword, like type "UNDEF" or "UNKNOWN" and use it in DECLARE part
4) it is already declared variable in declare block, so there is not need to do nothing (they don't have @ before variable name in function body) 

My proposal is to move declaration of each of 3 types of @variables above to declare part of PostgreSQL function, after that remove @ in each variable usage in PostgreSQL converted function.
If you prefer, you can put some warning or text comment in declare part, that mention that user must manually set proper types of variables and to replace for example:
```
DECLARE
a INT
b CHAR
--you must manually set proper variable types of variables below
firstvariable "UNKNOWN"
secondvariable "UNKNOWN"
BEGIN
a: =
firstvalue :=
...(and so on, but @ is removed in function code).
```
I think that such declaration part will speed-up very much procedure/function conversion, also, in the future, checks for assignment type or guessing of result type based on query and context of columns
(or some function return type, if it used - for example, if you call some query that execute some string function, you already know it't return type, based on function prototype) and so on. 

Another example before conversion (there is nothing to say, different from above example, only to explain things better):
```
| F_ArtCounterGetNextNumber | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` FUNCTION `F_ArtCounterGetNextNumber`(HospitalInfoIDArg int(11), CounterTypeArg int(11), CounterYearArg int(11)) RETURNS int(11)
BEGIN

  UPDATE ivf_counters c
  SET c.CounterValue = c.CounterValue + 1
  WHERE c.HospitalInfoID = HospitalInfoIDArg
  AND c.CounterType = CounterTypeArg
  AND c.CounterYear = CounterYearArg;

  SET @CounterID := 0;
  SET @NextNum := 1;

  SELECT
    c.CounterID,
    c.CounterValue INTO @CounterID, @NextNum
  FROM ivf_counters c
  WHERE c.HospitalInfoID = HospitalInfoIDArg
  AND c.CounterType = CounterTypeArg
  AND c.CounterYear = CounterYearArg;

  IF (@CounterID = 0) THEN
    INSERT INTO ivf_counters (CounterValue, HospitalInfoID, CounterType, CounterYear)
      VALUES (1, HospitalInfoIDArg, CounterTypeArg, CounterYearArg);
    RETURN 1;
  END IF;



  RETURN @NextNum;

END | utf8                 | utf8_general_ci      | utf8_general_ci    |
```
And right now, after conversion:
```
CREATE OR REPLACE FUNCTION f_artcountergetnextnumber (HospitalInfoIDArg bigint, CounterTypeArg bigint, CounterYearArg bigint) RETURNS integer AS $body$
BEGIN

  UPDATE ivf_counters c
  SET c.CounterValue = c.CounterValue + 1
  WHERE c.HospitalInfoID = HospitalInfoIDArg
  AND c.CounterType = CounterTypeArg
  AND c.CounterYear = CounterYearArg;

  SET @CounterID := 0;
  SET @NextNum := 1;

  SELECT
    c.CounterID,
    c.CounterValue INTO @CounterID, @NextNum
  FROM ivf_counters c
  WHERE c.HospitalInfoID = HospitalInfoIDArg
  AND c.CounterType = CounterTypeArg
  AND c.CounterYear = CounterYearArg;

  IF (@CounterID = 0) THEN
    INSERT INTO ivf_counters(CounterValue, HospitalInfoID, CounterType, CounterYear)
      VALUES (1, HospitalInfoIDArg, CounterTypeArg, CounterYearArg);
    RETURN 1;
  END IF;



  RETURN @NextNum;

END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION f_artcountergetnextnumber (HospitalInfoIDArg bigint, CounterTypeArg bigint, CounterYearArg bigint) FROM PUBLIC;
```
I think that such declaration (only once, because it is checked if it already exists or not) in declare part and simply setting to some dummy type as "UNDEF" or "UNKNOWN" will make very easy porting of MySQL stored procedures and functions and also, it will not block future enhancements, if possible at all. It is always easy to see what is declared in declare part and set it's proper value once, that manually edit and check whole stored procedures and functions.

**Comment by krasandr** [comment]
Another example, more complex, with mix of declared and @variables and some of them are set by return type of result of assignment:

| P_WAppointmentTempGenerator | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` PROCEDURE `P_WAppointmentTempGenerator`(IN p_WorkAppMetaID int)
BEGIN


  DECLARE p_From datetime;
  DECLARE p_To datetime;
  DECLARE p_Step int;
  DECLARE p_AppID int;
  DECLARE p_work_date date;


  DELETE
    FROM w_appointment_cache
  WHERE WorkAppMetaID IN (p_WorkAppMetaID);

  
  INSERT INTO w_appointment_cache (WorkAppMetaID, WCacheFromTime, WCacheToTime,
  IsFree, Duration, WAppointmentPatientID, WorkDate)
    SELECT
      p.WorkAppMetaID,
      p.WAppointmentFromDateTime,
      p.WAppointmentToDateTime,
      0 AS IsFree,
      p.WDuration,
      p.WAppointmentPatientID,
      p.WorkDate
    FROM v_wappointment_patients p
    WHERE p.WorkAppMetaID IN (p_WorkAppMetaID)
    AND p.IsCanceled = 0;


  
  SELECT
    m.WorkFromDateTime,
    m.WorkToDateTime,
    m.WorkStep,
    m.WorkAppMetaID,
    m.WorkDate INTO p_From, p_To, p_Step, p_AppID, p_work_date
  FROM v_wappointment_meta m
  WHERE m.WorkAppMetaID = p_WorkAppMetaID;


  WHILE (p_From <= p_To) DO

    SET @p_termp := DATE_ADD(p_From, INTERVAL p_Step MINUTE);

    SET @max_v := (SELECT
        MAX(p.WCacheToTime)
      FROM w_appointment_cache p
      WHERE p.WorkAppMetaID = p_AppID
      AND p.WCacheFromTime >= p_From
      AND p.WCacheFromTime < @p_termp);

    
    


    IF @max_v IS NULL THEN
      INSERT INTO w_appointment_cache (WorkAppMetaID, WCacheFromTime, WCacheToTime,
      IsFree, Duration, WAppointmentPatientID, WorkDate)
        VALUE (p_AppID, p_From, @p_termp, 1, p_Step, NULL, p_work_date);
    END IF;

    SET p_From = IFNULL(@max_v, DATE_ADD(@p_termp, INTERVAL 0 MINUTE));

  END WHILE;




END | utf8                 | utf8_general_ci      | utf8_general_ci    |

**Comment by krasandr** [comment]
If you accept such proposal, I think that result after conversion will be something like this (for last example procedure):

CREATE OR REPLACE FUNCTION p_wappointmenttempgenerator (IN p_WorkAppMetaID integer) RETURNS VOID AS $body$
DECLARE

p_From timestamp without time zone;
p_To timestamp without time zone;
p_Step integer;
p_AppID integer;
p_work_date date;
--added to declare part, must be set to proper type
p_termp "FIXME"
--added to declare part, must be set to proper type
max_v "FIXME"
BEGIN


  




  DELETE
    FROM w_appointment_cache
  WHERE WorkAppMetaID ,IN (p_WorkAppMetaID);

  
  INSERT INTO w_appointment_cache(WorkAppMetaID, WCacheFromTime, WCacheToTime,
  IsFree, Duration, WAppointmentPatientID, WorkDate)
    SELECT
      p.WorkAppMetaID,
      p.WAppointmentFromDateTime,
      p.WAppointmentToDateTime,
      0 AS IsFree,
      p.WDuration,
      p.WAppointmentPatientID,
      p.WorkDate
    FROM v_wappointment_patients p
    WHERE p.WorkAppMetaID ,IN (p_WorkAppMetaID)
    AND p.IsCanceled = 0;


  
  SELECT
    m.WorkFromDateTime,
    m.WorkToDateTime,
    m.WorkStep,
    m.WorkAppMetaID,
    m.WorkDate INTO p_From, p_To, p_Step, p_AppID, p_work_date
  FROM v_wappointment_meta m
  WHERE m.WorkAppMetaID = p_WorkAppMetaID;


  WHILE(p_From <= p_To) LOOP

    SET p_termp := (p_From)::timestamp + interval 'p_Step MINUTE';

    SET max_v := (SELECT
        MAX(p.WCacheToTime)
      FROM w_appointment_cache p
      WHERE p.WorkAppMetaID = p_AppID
      AND p.WCacheFromTime >= p_From
      AND p.WCacheFromTime < p_termp);

    
    


    IF max_v IS NULL THEN
      INSERT INTO w_appointment_cache(WorkAppMetaID, WCacheFromTime, WCacheToTime,
      IsFree, Duration, WAppointmentPatientID, WorkDate)
        VALUE(p_AppID, p_From, p_termp, 1, p_Step, NULL, p_work_date);
    END IF;

    p_From := COALESCE(max_v, (p_termp)::timestamp + interval '0 MINUTE');

  END LOOP;




END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION p_wappointmenttempgenerator (IN p_WorkAppMetaID integer) FROM PUBLIC;

**Comment by darold** [comment]
Thanks for all these explanations, I will follow your proposal to translate @variables. I think I will be able to work on it next weekend.

**Comment by krasandr** [comment]
I forgot one additional detail - it is possible in stored procedures (maybe in functions also) in MySQL to set some per-session configuration variables (like PostgreSQL GUC), for example:

 @@SESSION.unique_checks := 0;
  @@SESSION.foreign_key_checks := 0;
  @@sql_log_bin := 0;
  @@sql_log_bin := 0;

They must not be translated, because implementation in PostgreSQL is different, so they must be replaced manually. I mention them, because they are always with two @ in beginning of @variable,
so there must be check, to distinguish them based on number of @ before variable name, so they must remain intact.

**Comment by krasandr** [comment]
@variable can be found also in trigger (in trigger function), but all is the same as ordinary stored procedures and functions, so I post here two examples of such triggers:

| dp_before_insert | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` TRIGGER dp_before_insert
	BEFORE INSERT
	ON department_patients
	FOR EACH ROW
BEGIN
  SET @depPatID := new.DepartmentPationID;
  SET @clPathID := new.OutClinicPathID;

  SET new.PatientMinStay := IFNULL((SELECT
      cp.ClinicPathMinStay
    FROM clinic_paths cp
    WHERE cp.ClinicPathID = @clPathID), 0);



END | utf8                 | utf8_general_ci      | utf8_general_ci    | 2017-10-04 16:24:31.94 |

and example two:

| department_requests_after_del_tr | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`localhost` TRIGGER department_requests_after_del_tr
	AFTER DELETE
	ON department_requests
	FOR EACH ROW
BEGIN
set @depID:=(select  pm1.DepartmentID
from patient_movements pm1,
patient_movements pm2
where pm1.DepartmentPationID=pm2.DepartmentPationID
and pm1.DepartmentPationID=old.DepartmentPationID
group by pm1.DepartmentPationID,pm1.PatientMovementID
having pm1.PatientMovementID=max(pm2.PatientMovementID));

update departments d
set d.DepartmentInCounter=d.DepartmentInCounter+1
where d.DepartmentID=@depID;

update const_patient_statuses st
set st.PSInConsultationCounter=st.PSInConsultationCounter+1
where st.PatientStatusID=old.PatientStatusID;

update doctors d
set d.DoctorCounter=d.DoctorCounter+1
where d.DoctorID=old.DoctorID;

END | utf8                 | utf8_general_ci      | utf8_general_ci    | 2017-10-04 16:24:32.45 |

right now, after conversion with ora2pg, result is:

DROP TRIGGER IF EXISTS dp_before_insert ON department_patients CASCADE;
CREATE OR REPLACE FUNCTION trigger_fct_dp_before_insert() RETURNS trigger AS $BODY$
BEGIN
  SET @depPatID := new.DepartmentPationID;
  SET @clPathID := new.OutClinicPathID;

  SET new.PatientMinStay := COALESCE((SELECT
      cp.ClinicPathMinStay
    FROM clinic_paths cp
    WHERE cp.ClinicPathID = @clPathID), 0);



RETURN NEW;
END
$BODY$
 LANGUAGE 'plpgsql';

CREATE TRIGGER dp_before_insert
	BEFORE INSERT ON department_patients FOR EACH ROW
	EXECUTE PROCEDURE trigger_fct_dp_before_insert();

and for second example:

DROP TRIGGER IF EXISTS department_requests_after_del_tr ON department_requests CASCADE;
CREATE OR REPLACE FUNCTION trigger_fct_department_requests_after_del_tr() RETURNS trigger AS $BODY$
BEGIN
set @depID:=(select  pm1.DepartmentID
from patient_movements pm1,
patient_movements pm2
where pm1.DepartmentPationID=pm2.DepartmentPationID
and pm1.DepartmentPationID=old.DepartmentPationID
group by pm1.DepartmentPationID,pm1.PatientMovementID
having pm1.PatientMovementID=max(pm2.PatientMovementID));

update departments d
set d.DepartmentInCounter=d.DepartmentInCounter+1
where d.DepartmentID=@depID;

update const_patient_statuses st
set st.PSInConsultationCounter=st.PSInConsultationCounter+1
where st.PatientStatusID=old.PatientStatusID;

update doctors d
set d.DoctorCounter=d.DoctorCounter+1
where d.DoctorID=old.DoctorID;

RETURN OLD;
END
$BODY$
 LANGUAGE 'plpgsql';

CREATE TRIGGER department_requests_after_del_tr
	AFTER DELETE ON department_requests FOR EACH ROW
	EXECUTE PROCEDURE trigger_fct_department_requests_after_del_tr();

**Comment by krasandr** [comment]
I write to you in comment to commit message, but last commit broke all MySQL/MariaDB triggers, words "IF EXISTS" now appears twice, before they appeared only once:

DROP TRIGGER IF EXISTS IF EXISTS dp_before_insert ON department_patients CASCADE;
CREATE OR REPLACE FUNCTION trigger_fct_dp_before_insert() RETURNS trigger AS $BODY$
BEGIN
  SET @depPatID := new.DepartmentPationID;
  SET @clPathID := new.OutClinicPathID;

  SET new.PatientMinStay := COALESCE((SELECT
      cp.ClinicPathMinStay
    FROM clinic_paths cp
    WHERE cp.ClinicPathID = @clPathID), 0);



RETURN NEW;
END
$BODY$
 LANGUAGE 'plpgsql';

CREATE TRIGGER dp_before_insert
<------>BEFORE INSERT ON department_patients FOR EACH ROW
<------>EXECUTE PROCEDURE trigger_fct_dp_before_insert();

**Comment by darold** [comment]
Fixed sorry for the inconvenient.

**Comment by krasandr** [comment]
Yes, problem with trigger is fixed, thank you.

What you think about proposal for MySQL @variables? Maybe it will be more easy, if conversion is made incremental, step by step. In first step, it can be for example, if @variable is found in stored procedure or function (or trigger function), but not @@config_variable, only to declare it in declare block once, without proper type, for example with type of FIXME (or something else) and only to remove @ from variable usage in begin block. Also, it can be, (if you or someone other don't like something about conversion), to be added in mysql configuration part of ora2pg.conf , for example DECLARE_MYSQL_VARIABLES or something else, that can be set or unset by default.

I think that, in the future in the next steps, it can be used PostgreSQL catalog in the next pass (or before final writing to output file), so getting type of table.column or return type of calling or some builtin PostgreSQL function, or again from catalog, or regex to parse type of assignment, to be able to find right type of variable and replace FIXME with it in later steps, can be done. But it will not work for INPUT/OUTPUT type check of UDF, because they will not exists before creation time, unless they are created only as empty prototype, to be able to get via SQL from catalog, signature of function as black box (if some @variable get/set value from INPUT/OUTPUT type) - but this can be unneeded or more easy done with parsing original DDL and getting proper type for UDF from there.

I know, that you don't cache metadata for converted objects/schema before conversion, but if you have, for example, live connection to PostgreSQL after schema creation, maybe it will be not very hard (or slow) to use it, so to be able to get needed information dynamically, as last pass, maybe again with some configuration variable, something DYNAMIC_TYPE_CHECK - it can be used also for casting to string in string functions like L/R PAD, CONCAT and so on?

**Comment by darold** [comment]
Commit  8103254 adds this feature. There might be some adjustment following your feedback.

**Comment by krasandr** [comment]
There is multiple times declaration of the same variable in trigger function, for example before conversion:

| department_requests_after_upd_tr | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`localhost` TRIGGER department_requests_after_upd_tr
	AFTER UPDATE
	ON department_requests
	FOR EACH ROW
BEGIN
set @depID:=(select  pm1.DepartmentID
from patient_movements pm1,
patient_movements pm2
where pm1.DepartmentPationID=pm2.DepartmentPationID
and pm1.DepartmentPationID=new.DepartmentPationID
group by pm1.DepartmentPationID,pm1.PatientMovementID
having pm1.PatientMovementID=max(pm2.PatientMovementID));

update departments d
set d.DepartmentInCounter=d.DepartmentInCounter+1
where d.DepartmentID=@depID;

update const_patient_statuses st
set st.PSInConsultationCounter=st.PSInConsultationCounter+1
where st.PatientStatusID=new.PatientStatusID
or st.PatientStatusID=old.PatientStatusID;

update doctors d
set d.DoctorCounter=d.DoctorCounter+1
where d.DoctorID=new.DoctorID;

END | utf8                 | utf8_general_ci      | utf8_general_ci    | 2017-10-04 16:24:32.39 |

After conversion, there is three times declaration of @depID:

DROP TRIGGER IF EXISTS department_requests_after_upd_tr ON department_requests CASCADE;
CREATE OR REPLACE FUNCTION trigger_fct_department_requests_after_upd_tr() RETURNS trigger AS $BODY$
DECLARE
depID integer;
depID integer;
depID integer;

BEGIN
SET depID := (select  pm1.DepartmentID
from patient_movements pm1,
patient_movements pm2
where pm1.DepartmentPationID=pm2.DepartmentPationID
and pm1.DepartmentPationID=new.DepartmentPationID
group by pm1.DepartmentPationID,pm1.PatientMovementID
having pm1.PatientMovementID=max(pm2.PatientMovementID));

update departments d
set d.DepartmentInCounter=d.DepartmentInCounter+1
where d.DepartmentID=depID;

update const_patient_statuses st
set st.PSInConsultationCounter=st.PSInConsultationCounter+1
where st.PatientStatusID=new.PatientStatusID
or st.PatientStatusID=old.PatientStatusID;

update doctors d
set d.DoctorCounter=d.DoctorCounter+1
where d.DoctorID=new.DoctorID;

RETURN NEW;
END
$BODY$
 LANGUAGE 'plpgsql';

CREATE TRIGGER department_requests_after_upd_tr
<------>AFTER UPDATE ON department_requests FOR EACH ROW
<------>EXECUTE PROCEDURE trigger_fct_department_requests_after_upd_tr();

**Comment by darold** [comment]
Not tested but commit b33d337 might solved duplicate declaration of local variables.

**Comment by krasandr** [comment]
It solves this problem, thank you.
But there is another problem - if you have a @variable in IF(a,b,c) code block or in SELECT ... INTO @variables, they are not converted. Seems it is related to the order of evaluation of expressions with regex.

**Comment by krasandr** [comment]
There is problem with SET and := ( must not contain ":" )
For example, before conversion:

| f_hadisgetstatus | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` FUNCTION `f_hadisgetstatus`(patID int, inDatTime datetime, evType varchar(10)) RETURNS int(11)
BEGIN

  IF (inDatTime < date ('2015-04-01')) THEN
    RETURN -1;
  END IF;

  SET @EventType := evType;
  SET @StamEvDateTime := inDatTime;
  SET @patientID_1 := patID;

  RETURN (SELECT
      IF(COUNT(*) > 0, 1, 0) c
    FROM patients p
      INNER JOIN hadis h
        ON p.PatientID = h.PatientID
    WHERE p.PatientID = @patientID_1
    AND h.EventTimeStamp
    BETWEEN DATE_ADD(@StamEvDateTime, INTERVAL -4 HOUR) AND
    DATE_ADD(@StamEvDateTime, INTERVAL 4 HOUR)
    AND h.EventType = @EventType);

END | utf8                 | utf8_general_ci      | latin1_swedish_ci  |

With ora2pg, it is converted right, but with one mistake - PostreSQL don't accept := in SET variables
(so, I must delete ":" from assignment, after that it is O.K.):

CREATE OR REPLACE FUNCTION f_hadisgetstatus (patID integer, inDatTime timestamp without time zone, evType varchar(10)) RETURNS integer AS $body$
DECLARE
EventType integer;
StamEvDateTime timestamp;
patientID_1 integer;

BEGIN
  
  IF (inDatTime < date('2015-04-01')) THEN
    RETURN -1;
  END IF;
  
  SET EventType = evType;
  SET StamEvDateTime = inDatTime;
  SET patientID_1 = patID;
    
  RETURN(SELECT
      CASE WHEN COUNT(*) > 0 THEN 1 ELSE  0 END  c
    FROM patients p
      INNER JOIN hadis h
        ON p.PatientID = h.PatientID
    WHERE p.PatientID = patientID_1
    AND h.EventTimeStamp
    BETWEEN(StamEvDateTime)::timestamp + -4*interval '1 HOUR' AND (StamEvDateTime)::timestamp + 4*interval '1 HOUR'
    AND h.EventType = EventType);

END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION f_hadisgetstatus (patID integer, inDatTime timestamp without time zone, evType varchar(10)) FROM PUBLIC;

**Comment by krasandr** [comment]
Also, there is sometimes assignments, that after conversion contains "::=" in SET, that is also not accepted by PostgreSQL. For example, before conversion: 

| p_s_month_avg_min_productscalc | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` PROCEDURE `p_s_month_avg_min_productscalc`(IN immediatelyCalcul TINYINT)
BEGIN

set @frDate:= null; 
set @toDate:=DATE(now());
set @backVal:=(select s.GlobalSettingValue from global_settings s where s.GlobalSettingID=26);
set @frDate:=DATE_ADD(date(DATE_FORMAT(@toDate,'%Y-%m-%01')), interval - @backVal month);
set @toDate:=DATE_ADD(date(DATE_FORMAT(@toDate,'%Y-%m-%01')), interval - 1 day);
set @chekDate:=DATE_ADD(date(DATE_FORMAT(now(),'%Y-%m-%01')), interval - 1 day);

if(immediatelyCalcul=1 or  DATE(@toDate)=date(@chekDate))
then
  start transaction;
set @frDate:= null; 
set @toDate:=DATE(now());
set @backVal:=(select s.GlobalSettingValue from global_settings s where s.GlobalSettingID=26);
set @frDate:=DATE_ADD(date(DATE_FORMAT(@toDate,'%Y-%m-%01')), interval - @backVal month);
set @toDate:=DATE_ADD(date(DATE_FORMAT(@toDate,'%Y-%m-%01')), interval - 1 day);
set @chekDate:=DATE_ADD(date(DATE_FORMAT(now(),'%Y-%m-%01')), interval - 1 day);
DROP TEMPORARY TABLE IF EXISTS t_min;
CREATE temporary TABLE t_min
select t.ProductID,t.StoreID, sum(q)/@backVal qMonthAVG,sum(q)/@backVal as qForSupplay
from
(
SELECT sl.ProductID,sd.StoreID, SUM(sdd.Quantity) AS q
  FROM s_doc_details sdd
  INNER JOIN s_lots sl ON sdd.SLotID = sl.SLotID
  INNER JOIN s_documents sd ON sdd.SDocID = sd.SDocID
  INNER JOIN s_document_types sdt ON sd.SDocTypeID = sdt.SDocTypeID
  WHERE sdt.OrderOrSupply=-1
  AND sd.SDocDate BETWEEN date(@frDate) and date(@toDate)
  GROUP BY sl.ProductID, sd.StoreID
)as t 
group by ProductID,t.StoreID; 

UPDATE t_min m,
  s_products p,
  s_const_product_categories cat
  set m.qForSupplay=round(m.qForSupplay+m.qForSupplay*cat.AdditionalPercentMinQuan/100,2)
  WHERE
  m.ProductID=p.ProductID
  and cat.ProductCategoryID=p.ProductCategoryID;

DELETE from s_st_products_month_avg_costs;
insert into s_st_products_month_avg_costs(ProductID,StoreID,St_productCost)
SELECT t.ProductID,t.StoreID,qMonthAVG  from t_min t;
 insert into events_log(EventLogTime,EventLogDescription) values (now(),'Average actualization for month...');

DELETE from s_products_stores_quantity;
insert into s_products_stores_quantity(ProductID,StoreID,Quantity)
SELECT t.ProductID,t.StoreID,qForSupplay  from t_min t;
insert into events_log(EventLogTime,EventLogDescription) values (now(),'Average actualization for month...');

COMMIT;
end if;

END | utf8                 | utf8_general_ci      | latin1_swedish_ci  |

After conversion with Ora2Pg (actual git version from today), there is "::=" assignments, also there is @values, that are not converted to declare and still have "@" in it's name:

CREATE OR REPLACE FUNCTION p_s_month_avg_min_productscalc (IN immediatelyCalcul smallint) RETURNS VOID AS $body$
BEGIN

@frDate::= null;.
@toDate::=DATE(CURRENT_TIMESTAMP);
@backVal::=(select s.GlobalSettingValue from global_settings s where s.GlobalSettingID=26);
@frDate::=DATE_ADD(date(DATE_FORMAT(@toDate,'%Y-%m-%01')), interval - @backVal month);
@toDate::=DATE_ADD(date(DATE_FORMAT(@toDate,'%Y-%m-%01')), interval - 1 day);
@chekDate::=DATE_ADD(date(DATE_FORMAT(CURRENT_TIMESTAMP,'%Y-%m-%01')), interval - 1 day);

if (immediatelyCalcul=1 or  DATE(@toDate)=date(@chekDate))
then
 ---- START TRANSACTION;
@frDate::= null;.
@toDate::=DATE(CURRENT_TIMESTAMP);
@backVal::=(select s.GlobalSettingValue from global_settings s where s.GlobalSettingID=26);
@frDate::=DATE_ADD(date(DATE_FORMAT(@toDate,'%Y-%m-%01')), interval - @backVal month);
@toDate::=DATE_ADD(date(DATE_FORMAT(@toDate,'%Y-%m-%01')), interval - 1 day);
@chekDate::=DATE_ADD(date(DATE_FORMAT(CURRENT_TIMESTAMP,'%Y-%m-%01')), interval - 1 day);
DROP TEMPORARY TABLE IF EXISTS t_min;
CREATE temporary TABLE t_min
select t.ProductID,t.StoreID, sum(q)/@backVal qMonthAVG,sum(q)/@backVal as qForSupplay
from (
SELECT sl.ProductID,sd.StoreID, SUM(sdd.Quantity) AS q
  FROM s_doc_details sdd
  INNER JOIN s_lots sl ON sdd.SLotID = sl.SLotID
  INNER JOIN s_documents sd ON sdd.SDocID = sd.SDocID
  INNER JOIN s_document_types sdt ON sd.SDocTypeID = sdt.SDocTypeID
  WHERE sdt.OrderOrSupply=-1
  AND sd.SDocDate BETWEEN date(@frDate) and date(@toDate)
  GROUP BY sl.ProductID, sd.StoreID
)as t.
group by ProductID,t.StoreID;.

UPDATE t_min m,
  s_products p,
  s_const_product_categories cat
  set m.qForSupplay=round((m.qForSupplay+m.qForSupplay*cat.AdditionalPercentMinQuan/100)::numeric,2)
  WHERE
  m.ProductID=p.ProductID
  and cat.ProductCategoryID=p.ProductCategoryID;

DELETE from s_st_products_month_avg_costs;
insert into s_st_products_month_avg_costs(ProductID,StoreID,St_productCost)
SELECT t.ProductID,t.StoreID,qMonthAVG  from t_min t;
 insert into events_log(EventLogTime,EventLogDescription) values (CURRENT_TIMESTAMP,'Average actualization for month...');

DELETE from s_products_stores_quantity;
insert into s_products_stores_quantity(ProductID,StoreID,Quantity)
SELECT t.ProductID,t.StoreID,qForSupplay  from t_min t;
insert into events_log(EventLogTime,EventLogDescription) values (CURRENT_TIMESTAMP,'Average actualization for month...');

-- -- COMMIT;
end if;

END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION p_s_month_avg_min_productscalc (IN immediatelyCalcul smallint) FROM PUBLIC;

**Comment by darold** [comment]
Commit 6e6b673 might fix remaining translation issues for @variable reported in the two previous note.

**Comment by darold** [comment]
There is still some not translated variable call in the files you sent to me, I will fix them tomorrow.

**Comment by krasandr** [comment]
Previous two notes are fixed, but there is one new bug after some of last fixes - there is comma before IN, every time, for example before conversion:
| f_getnextfindocnumber | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` FUNCTION `f_getnextfindocnumber`(FinDocType int(11), HospitalInfoID int(11)) RETURNS bigint(20)
BEGIN

  IF (FinDocType IN (4, 5)) THEN
    RETURN (SELECT
        MAX(t.NextNum)
      FROM (SELECT
          IFNULL(MAX(fd.FDocNum), 0) + 1 AS NextNum
        FROM f_docs fd
        WHERE fd.FDocTypeID IN (4, 5)
        AND fd.HospitalInfoID = HospitalInfoID

        UNION ALL

        SELECT
          IFNULL(MAX(piv.InvoiceNumber), 0) + 1 AS NextNum
        FROM patient_invoices piv
        WHERE piv.HospitalInfoID = HospitalInfoID
        AND piv.InvoiceTypeID IN (1, 2, 3)) AS t);
  END IF;

  RETURN (SELECT
      IFNULL(MAX(fd.FDocNum), 0) + 1
    FROM f_docs fd
    WHERE fd.FDocTypeID = FinDocType
    AND fd.HospitalInfoID = HospitalInfoID);

END | utf8                 | utf8_general_ci      | latin1_swedish_ci  |

And after conversion, see ",IN", three times in this function (if I remove comma, function is created successful):

CREATE OR REPLACE FUNCTION f_getnextfindocnumber (FinDocType bigint, HospitalInfoID bigint) RETURNS bigint AS $body$
BEGIN

  IF (FinDocType ,IN (4, 5)) THEN
    RETURN(SELECT
        MAX(t.NextNum)
      FROM (SELECT
          COALESCE(MAX(fd.FDocNum), 0) + 1 AS NextNum
        FROM f_docs fd
        WHERE fd.FDocTypeID ,IN (4, 5)
        AND fd.HospitalInfoID = HospitalInfoID
        
        UNION ALL
        
        SELECT
          COALESCE(MAX(piv.InvoiceNumber), 0) + 1 AS NextNum
        FROM patient_invoices piv
        WHERE piv.HospitalInfoID = HospitalInfoID
        AND piv.InvoiceTypeID ,IN (1, 2, 3)) AS t);
  END IF;

  RETURN(SELECT
      COALESCE(MAX(fd.FDocNum), 0) + 1
    FROM f_docs fd
    WHERE fd.FDocTypeID = FinDocType
    AND fd.HospitalInfoID = HospitalInfoID);

END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION f_getnextfindocnumber (FinDocType bigint, HospitalInfoID bigint) FROM PUBLIC;

**Comment by krasandr** [comment]
A long, but simple function, that covers maybe all problems, that are not yet resolved - there is three different problems - comma before IN -  ",IN" or "NOT ,IN" and so on, @variables that are not converted and that is found in SELECT ... INTO @variable list, and RETURN @variable or calling function with @variable parameter.
Before conversion:
| f_artcompletiondate | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` FUNCTION `f_artcompletiondate`(artID int) RETURNS date
BEGIN


  
  
  
  SET @IvfArtID = artID;
  
  SET @ArtStatusID := (SELECT
      ars.IvfArtStatusID 
    FROM ivf_const_art_statuses ars
    WHERE ars.IvfArtStatusID = F_ArtStatus(@IvfArtID, 0)
    AND ars.IsEnded = 1);

  
  IF (@ArtStatusID IS NULL)
    THEN
    RETURN NULL;
  END IF;

  
  SELECT
    s.IvfStimID,
    p.IvfPunctionID,
    a.IvfArtFormID,
    DATE(p.IvfPunctionDateTime) AS puncDate,
    a.IvfArtStartDateTime INTO @stimID, @puncID, @IvfFormID, @puncDate, @artStartDate

  FROM ivf_arts a
    INNER JOIN ivf_stimulations s 
      ON (a.IvfArtID = s.IvfArtID)
    LEFT JOIN ivf_punctions p
      ON s.IvfStimID = p.IvfStimID
  WHERE s.IvfArtID = @IvfArtID;


  
  
  
  
  
  
  
  
  
  


  
  
  IF (@ArtStatusID IN (5, 6))
    THEN
    RETURN F_GetStimDateCompleted(@stimID);
  END IF;


  
  IF (@ArtStatusID = 8)
    THEN
    RETURN F_GetPuncDateCompleted(@puncID);
  END IF;

  
  IF (@ArtStatusID = 9)
    THEN
    RETURN DATE(@artStartDate);
  END IF;



  
  
  
  IF (@ArtStatusID IN (10, 11, 12))
    THEN

    RETURN (SELECT
        MAX(v.CalculEndDevDate)
      FROM v_ocells v
      WHERE v.IvfArtID = @IvfArtID);
  END IF;

  
  IF (@ArtStatusID = 13)
    THEN
    RETURN (SELECT
        MAX(DATE(c.IvfInsFerDateTime))
      FROM ivf_ins_fertilizations c,
           ivf_ins_fertilizations_ivf_s_cells fc
      WHERE c.IvfStimID = @stimID
      AND fc.IvfInsFerID = c.IvfInsFerID);
  END IF;

  
  IF (@ArtStatusID = 15)
    THEN
    RETURN @puncDate;
  END IF;

  
  IF (@ArtStatusID = 16)
    THEN
    RETURN (DATE_ADD(DATE(@artStartDate), INTERVAL 10 DAY));
  END IF;















  IF (@ArtStatusID = 5
    || @ArtStatusID = 6)
    THEN
    RETURN (SELECT
        IFNULL(IFNULL(MAX(sd.IvfStimDetailDate), s.IvfStimPregnylDateTime), s.IvfStimDate) AS dat

      FROM ivf_stimulations s
        LEFT JOIN (ivf_stimulations_details sd)
          ON (s.IvfStimID = sd.IvfStimID)
      WHERE s.IvfArtID = @IvfArtID);

  END IF;

  IF (@ArtStatusID = 8)
    THEN
    RETURN (SELECT
        pu.IvfPunctionDateTime AS dat

      FROM ivf_stimulations s,
           ivf_punctions pu
      WHERE s.IvfStimID = pu.IvfStimID
      AND s.IvfArtID = @IvfArtID);

  END IF;

  
  IF (@ArtStatusID = 10)
    THEN
    RETURN (SELECT
        MAX(s.IvfTransferDateTime) AS dat

      FROM ivf_transfers s
      WHERE s.IvfArtID = @IvfArtID);

  END IF;

  
  IF (@ArtStatusID = 11)
    THEN
    RETURN (SELECT
        MAX(s.IvfFreezingDate) AS dat

      FROM ivf_cryo_eggs s,
           ivf_o_cells_copy o
      WHERE s.IvfOCellID = o.IvfOCellID
      AND o.IvfArtID = @IvfArtID);

  END IF;

  

  IF (@ArtStatusID = 12)
    THEN
    SET @freseDate := (SELECT
        MAX(s.IvfFreezingDate) AS dat

      FROM ivf_cryo_eggs s,
           ivf_o_cells_copy o
      WHERE s.IvfOCellID = o.IvfOCellID
      AND o.IvfArtID = @IvfArtID);

    SET @transDate := (SELECT
        s.IvfTransferDateTime AS dat

      FROM ivf_transfers s
      WHERE s.IvfArtID = @IvfArtID);

    RETURN GREATEST(@freseDate, @transDate);

  END IF;

  RETURN NULL;
END | utf8                 | utf8_general_ci      | latin1_swedish_ci  |

After conversion, with all problems mention above, included:
CREATE OR REPLACE FUNCTION f_artcompletiondate (artID integer) RETURNS date AS $body$
DECLARE
ArtStatusID integer;
freseDate timestamp;
transDate timestamp;

BEGIN


  
  
  
  @IvfArtID := artID;
  
  SET ArtStatusID = (SELECT
      ars.IvfArtStatusID 
    FROM ivf_const_art_statuses ars
    WHERE ars.IvfArtStatusID = F_ArtStatus(@IvfArtID, 0)
    AND ars.IsEnded = 1);

  
  IF (ArtStatusID IS NULL)
    THEN
    RETURN NULL;
  END IF;

  
  SELECT
    s.IvfStimID,
    p.IvfPunctionID,
    a.IvfArtFormID,
    DATE(p.IvfPunctionDateTime) AS puncDate,
    a.IvfArtStartDateTime INTO @stimID, @puncID, @IvfFormID, @puncDate, @artStartDate

  FROM ivf_arts a
    INNER JOIN ivf_stimulations s 
      ON (a.IvfArtID = s.IvfArtID)
    LEFT JOIN ivf_punctions p
      ON s.IvfStimID = p.IvfStimID
  WHERE s.IvfArtID = @IvfArtID;


  
  
  
  
  
  
  
  
  
  


  
  
  IF (ArtStatusID ,IN (5, 6))
    THEN
    RETURN F_GetStimDateCompleted(@stimID);
  END IF;


  
  IF (ArtStatusID = 8)
    THEN
    RETURN F_GetPuncDateCompleted(@puncID);
  END IF;

  
  IF (ArtStatusID = 9)
    THEN
    RETURN DATE(@artStartDate);
  END IF;



  
  
  
  IF (ArtStatusID ,IN (10, 11, 12))
    THEN

    RETURN(SELECT
        MAX(v.CalculEndDevDate)
      FROM v_ocells v
      WHERE v.IvfArtID = @IvfArtID);
  END IF;

  
  IF (ArtStatusID = 13)
    THEN
    RETURN(SELECT
        MAX(DATE(c.IvfInsFerDateTime))
      FROM ivf_ins_fertilizations c,
           ivf_ins_fertilizations_ivf_s_cells fc
      WHERE c.IvfStimID = @stimID
      AND fc.IvfInsFerID = c.IvfInsFerID);
  END IF;

  
  IF (ArtStatusID = 15)
    THEN
    RETURN @puncDate;
  END IF;

  
  IF (ArtStatusID = 16)
    THEN
    RETURN ((DATE(@artStartDate))::timestamp + 10*interval '1 DAY');
  END IF;















  IF (ArtStatusID = 5
     OR  ArtStatusID = 6)
    THEN
    RETURN(SELECT
        COALESCE(COALESCE(MAX(sd.IvfStimDetailDate), s.IvfStimPregnylDateTime), s.IvfStimDate) AS dat

      FROM ivf_stimulations s
        LEFT JOIN(ivf_stimulations_details sd)
          ON (s.IvfStimID = sd.IvfStimID)
      WHERE s.IvfArtID = @IvfArtID);

  END IF;

  IF (ArtStatusID = 8)
    THEN
    RETURN(SELECT
        pu.IvfPunctionDateTime AS dat

      FROM ivf_stimulations s,
           ivf_punctions pu
      WHERE s.IvfStimID = pu.IvfStimID
      AND s.IvfArtID = @IvfArtID);

  END IF;

  
  IF (ArtStatusID = 10)
    THEN
    RETURN(SELECT
        MAX(s.IvfTransferDateTime) AS dat

      FROM ivf_transfers s
      WHERE s.IvfArtID = @IvfArtID);

  END IF;

  
  IF (ArtStatusID = 11)
    THEN
    RETURN(SELECT
        MAX(s.IvfFreezingDate) AS dat

      FROM ivf_cryo_eggs s,
           ivf_o_cells_copy o
      WHERE s.IvfOCellID = o.IvfOCellID
      AND o.IvfArtID = @IvfArtID);

  END IF;

  

  IF (ArtStatusID = 12)
    THEN
    SET freseDate = (SELECT
        MAX(s.IvfFreezingDate) AS dat

      FROM ivf_cryo_eggs s,
           ivf_o_cells_copy o
      WHERE s.IvfOCellID = o.IvfOCellID
      AND o.IvfArtID = @IvfArtID);

    SET transDate = (SELECT
        s.IvfTransferDateTime AS dat

      FROM ivf_transfers s
      WHERE s.IvfArtID = @IvfArtID);

    RETURN GREATEST(freseDate, transDate);

  END IF;

  RETURN NULL;
END;
$body$
LANGUAGE PLPGSQL
SECURITY DEFINER
;
-- REVOKE ALL ON FUNCTION f_artcompletiondate (artID integer) FROM PUBLIC;

**Comment by darold** [comment]
Those translation issue might be fixed with last commit 104c399.

**Comment by krasandr** [comment]
Great work, thank you, all translations of @variable are correct. I will test some more examples and will close this big.
The only issue that I found is related to conversion of IF (COUNT(*) = 0, 0, 1) and IF (COUNT(*) > 0, 1, 0) - they are not converted to CASE ... WHEN ... THEN ... ELSE ..., see example below:

| f_artstatus | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` FUNCTION `f_artstatus`(p_IvfArtID int,
IsCancel int) RETURNS int(11)
BEGIN



  SET @IvfArtID := p_IvfArtID;
  SET @IsCancel := IsCancel;


  SELECT
    a.IvfArtFormID,
    a.IvfArtStatusID,
    arts.IsCanceled INTO @IvfFormID, @OldArtStatusID, @IsOldCanceled
  FROM ivf_arts a,
       ivf_const_art_statuses arts
  WHERE a.IvfArtID = @IvfArtID
  AND a.IvfArtStatusID = arts.IvfArtStatusID;

  
  IF (@OldArtStatusID = 16)
    THEN
    RETURN 16;
  END IF;


  
  SET @PriklichilaDonorka :=
  IFNULL((SELECT
      1 AS b
    FROM ivf_stimulations s
      INNER JOIN ivf_punctions p
        ON s.IvfStimID = p.IvfStimID
    WHERE s.IvfArtID = @IvfArtID
    AND NOT EXISTS (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = s.IvfArtID
      AND o.IvfOStatusID = 1)), 0);

  IF (@PriklichilaDonorka = 1
    AND @IvfFormID = 4)
    THEN
    RETURN 15;
  END IF;

  
  SET @IsFertiliz := IFNULL((SELECT
      IF(COUNT(*) = 0, 0, 1)
    FROM ivf_stimulations s,
         ivf_ins_fertilizations c,
         ivf_ins_fertilizations_ivf_s_cells fc
    WHERE s.IvfStimID = c.IvfStimID
    AND fc.IvfInsFerID = c.IvfInsFerID
    AND s.IvfArtID = @IvfArtID), 0);
  IF (@IsFertiliz = 1)
    THEN
    RETURN 13;
  END IF;

  
  SET @IsEmbrioDev := IFNULL((SELECT
      IF(COUNT(*) > 0, 1, 0)
    FROM v_ocells o
    WHERE o.IvfArtID = @IvfArtID
    AND o.IvfOStatusID = 1), 0);

  IF (@IsEmbrioDev = 1)
    THEN
    RETURN 7;
  END IF;


  
  SET @IsTransfer := (SELECT
      1
    FROM v_ocells v
    WHERE v.IvfArtID = @IvfArtID
    AND v.IvfTransferID IS NOT NULL
    LIMIT 1);

  
  SET @IsFRZEmbrions := (SELECT
      1
    FROM v_ocells v
    WHERE v.IvfArtID = @IvfArtID
    AND v.IvfCryoEggID IS NOT NULL
    LIMIT 1);

  
  IF (@IsTransfer = 1
    AND @IsFRZEmbrions = 1)
    THEN
    RETURN 12;
  END IF;

  
  IF (@IsFRZEmbrions = 1)
    THEN
    RETURN 11;
  END IF;


  
  IF (@IsTransfer = 1)
    THEN
    RETURN 10;
  END IF;

  
  IF (@IsCancel = 1
    AND @IvfFormID = 3)
    THEN
    RETURN 9;
  END IF;



  
  SET @IsCancelAfterPunc :=
  IFNULL((SELECT
      1 AS b
    FROM ivf_stimulations s,
         ivf_punctions p
    WHERE s.IvfStimID = p.IvfStimID
    AND s.IvfArtID = @IvfArtID
    AND NOT EXISTS (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = s.IvfArtID
      AND o.IvfOStatusID != 4)), 0);

  IF (@IsCancelAfterPunc = 1)
    THEN
    RETURN 8;
  END IF;



  
  IF (@IsOldCanceled = 1)
    THEN
    SET @IsCancel = IF(@IsCancel = 0, 1, 0);
  END IF;


  
  
  IF (@IsCancel = 1
    AND @IvfFormID = 5)
    THEN
    RETURN 5;
  END IF;


  
  IF (@IsCancel = 1
    AND @IvfFormID = 3)
    THEN
    RETURN 9;
  END IF;

  IF (@IsCancel = 1)
    THEN
    RETURN 6;
  END IF;

  SET @IsEmbrCancel := (SELECT
      
      
      IF(COUNT(*) > 0
      AND COUNT(*) = SUM(IF(iod.IvfOStatusID = 4, 1, 0)),
      1, 0) AS Result

    FROM ivf_embryo_protocols iep
      INNER JOIN ivf_ocells_developments iod
        ON iep.EmbryoProtocolID = iod.EmbryoProtocolID

    WHERE iep.IvfArtID = @IvfArtID);

  IF (@IsEmbrCancel = 1) THEN
    RETURN 9;
  END IF;

  
  SET @IsStim := IFNULL((SELECT
      1
    FROM ivf_stimulations s
    WHERE s.IvfArtID = @IvfArtID), 0);


  
  IF (@IsStim = 1
    AND @IvfFormID = 5)
    THEN
    RETURN 4;
  END IF;

  
  IF (@IsStim = 1)
    THEN
    RETURN 3;
  END IF;

  
  IF (@IvfFormID = 2)
    THEN
    SET @IsIschakvane := (SELECT
        1
      FROM v_ocells o
      WHERE o.IvfArtID = @IvfArtID
      LIMIT 1);

    IF (@IsIschakvane != 1)
      THEN
      RETURN 2;
    END IF;
  END IF;


  
  RETURN 1;

END | utf8                 | utf8_general_ci      | latin1_swedish_ci  |

After conversion, the two IF conditions with COUNT(*) are not converted, so they contain in problematic part original code untouched:

 SET IsFertiliz = COALESCE((SELECT
      IF (COUNT(*) = 0, 0, 1)
    FROM ivf_stimulations s,
         ivf_ins_fertilizations c,
         ivf_ins_fertilizations_ivf_s_cells fc
    WHERE s.IvfStimID = c.IvfStimID
    AND fc.IvfInsFerID = c.IvfInsFerID
    AND s.IvfArtID = IvfArtID), 0);
  IF (IsFertiliz = 1)
    THEN
    RETURN 13;
  END IF;
  
..
  SET IsEmbrioDev = COALESCE((SELECT
      IF (COUNT(*) > 0, 1, 0)
    FROM v_ocells o
    WHERE o.IvfArtID = IvfArtID
    AND o.IvfOStatusID = 1), 0);
    
  IF (IsEmbrioDev = 1)
    THEN
    RETURN 7;
  END IF;

**Comment by krasandr** [comment]
By the way, problem with comma in ",IN" still exists, but only in VIEW definitions.

**Comment by krasandr** [comment]
Also, another issue (not related to @variable), in the note above from 2 days ago that covers MySQL stored procedure p_s_month_avg_min_productscalc,
I see in yours code there is some translations in PLSQL.pm line 2154-2161 of DATE_ADD and DATE_FORMAT (i.e.
$str =~ s/\bDATE_FORMAT\(\s*(.*?)\s*,\s*('[^\(\)]+')\s*\)/_mysql_dateformat_to_pgsql($1, $2)/eigs;
)
but after conversion, these MySQL functions still exists, so there is no TO_CHAR or some other conversion of them:

frDate := DATE_ADD(date(DATE_FORMAT(toDate,'%Y-%m-%01')), interval - backVal month);
toDate := DATE_ADD(date(DATE_FORMAT(toDate,'%Y-%m-%01')), interval - 1 day);
chekDate := DATE_ADD(date(DATE_FORMAT(CURRENT_TIMESTAMP,'%Y-%m-%01')), interval - 1 day);

if (immediatelyCalcul=1 or  DATE(toDate)=date(chekDate))
then
 ---- START TRANSACTION;
frDate := null;.
toDate := DATE(CURRENT_TIMESTAMP);
backVal := (select s.GlobalSettingValue from global_settings s where s.GlobalSettingID=26);
frDate := DATE_ADD(date(DATE_FORMAT(toDate,'%Y-%m-%01')), interval - backVal month);
toDate := DATE_ADD(date(DATE_FORMAT(toDate,'%Y-%m-%01')), interval - 1 day);
chekDate := DATE_ADD(date(DATE_FORMAT(CURRENT_TIMESTAMP,'%Y-%m-%01')), interval - 1 day);

**Comment by krasandr** [comment]
Seems problem with DATE_FORMAT is related to too much syntax differences and additional regex check, that are needed.
For example, not translated functions calling are:
    CONCAT(DATE_FORMAT(FromDate, '%d-%m-%Y'), ' / ', DATE_FORMAT(
    CONCAT(DATE_FORMAT(FromDate, '%d-%m-%Y'), ' / ', DATE_FORMAT(
    CONCAT(DATE_FORMAT(FromDate, '%d-%m-%Y'), ' / ', DATE_FORMAT(
    DATE_FORMAT(f.IvfArtStartDateTime, '%d-%m-%Y') AS start_date,
    DATE_FORMAT(pu.IvfPunctionDateTime, '%d-%m-%Y') AS punctionDate,
    DATE_FORMAT(pu.IvfPunctionDateTime, '%H:%i') AS punctionTime,
    DATE_FORMAT(st.DonorPRM, '%d-%m-%Y') AS PatientPRM,
    DATE_FORMAT(trans.IvfTransferDateTime, '%d-%m-%Y') AS TransDate,
    DATE_FORMAT(trans.IvfTransferDateTime, '%H:%i') AS TransTime,
      DATE_FORMAT(p.PatientBirthDay, '%d.%m.%Y') AS PatientBirthDay,
          DATE_FORMAT(r.RestHolidayFromDate, '%d.%m.%Y')
          DATE_FORMAT(r.RestHolidayToDate, '%d.%m.%Y')
      CONCAT_WS(' ', 'Example1', DATE_FORMAT(p.PatientBirthDay, '%d.%m.%Y'), 'Example2', CAST(p.PatientAge AS char)) AS pD,
      DATE_FORMAT(d.AmbDocumentDate, '%d.%m.%Y') AS amDate,
      DATE_FORMAT(d.AmbDocumentDate, '%H:%i') AS amHour,
      DATE_FORMAT(p.PatientBirthDay, '%d-%m-%Y') AS PatientBirthDay,
      DATE_FORMAT(CURRENT_TIMESTAMP, '%d.%m.%Y') CurrDate,
      DATE_FORMAT(CURRENT_TIMESTAMP, '%H:%i') CurrTime,
      DATE_FORMAT(p.PatientBirthDay, '%d.%m.%Y') AS PatientBirthDay,
      DATE_FORMAT(InDateTime, '%d.%m.%Y') AS inDate,
      DATE_FORMAT(OutDateTime, '%d.%m.%Y') AS outDate,
      DATE_FORMAT(InDateTime, '%H:%i') AS inTime,
      DATE_FORMAT(OutDateTime, '%H:%i') AS outTime,
      DATE_FORMAT(CURRENT_TIMESTAMP, '%d.%m.%Y') CurrDate,
      DATE_FORMAT(CURRENT_TIMESTAMP, '%H:%i') CurrTime,
      DATE_FORMAT(dp.HistologicalResultDate, '%d.%m.%Y') AS HistologicalResultDate,
      DATE_FORMAT(dpe.PatientOperationDate, '%d.%m.%Y') AS PatientOperationDate,
        DATE_FORMAT(je.OperationEndDateTime, '%Y') AS oYear,
        DATE_FORMAT(je.OperationEndDateTime, '%d-%m-%Y') oDate,
        DATE_FORMAT(je.OperationStartDateTime, '%d-%m-%Y %H:%i') oStart,
        DATE_FORMAT(je.OperationEndDateTime, '%d-%m-%Y %H:%i') oEnd,
        DATE_FORMAT(je.OperationStartDateTime, '%H:%i') AS oStartTime,
        DATE_FORMAT(je.OperationEndDateTime, '%H:%i') AS oEndTime,
      DATE_FORMAT(p.PatientBirthDay, '%d-%m-%Y') AS PatientBirthDay,
      DATE_FORMAT(CURRENT_TIMESTAMP, '%d.%m.%Y') CurrDate,
      DATE_FORMAT(CURRENT_TIMESTAMP, '%H:%i') CurrTime,
      CAST(DATE_FORMAT(p.PatientBirthDay, '%d-%m-%Y') AS char) AS PatientBirthDay,
      DATE_FORMAT(CURRENT_TIMESTAMP, '%d.%m.%Y') CurrDate,
      DATE_FORMAT(CURRENT_TIMESTAMP, '%H:%i') CurrTime,
      CONCAT(DATE_FORMAT(o.GrStartDateTime, '%H:%i'), '  ',
      DATE_FORMAT(o.GrStartDateTime, '%d-%m-%Y') tim,
      CONCAT(DATE_FORMAT(GrStartDateTime, '%H:%i'), '-',
      DATE_FORMAT(GrEndDateTime, '%H:%i')) AS stEnd,
      CONCAT(DATE_FORMAT(fromDate, '%d-%m-%Y'), ' / ',
      DATE_FORMAT(toDate, '%d-%m-%Y')) AS fromDD;
frDate := DATE_ADD(date(DATE_FORMAT(toDate,'%Y-%m-%01')), interval - backVal month);
toDate := DATE_ADD(date(DATE_FORMAT(toDate,'%Y-%m-%01')), interval - 1 day);
chekDate := DATE_ADD(date(DATE_FORMAT(CURRENT_TIMESTAMP,'%Y-%m-%01')), interval - 1 day);
frDate := DATE_ADD(date(DATE_FORMAT(toDate,'%Y-%m-%01')), interval - backVal month);
toDate := DATE_ADD(date(DATE_FORMAT(toDate,'%Y-%m-%01')), interval - 1 day);
chekDate := DATE_ADD(date(DATE_FORMAT(CURRENT_TIMESTAMP,'%Y-%m-%01')), interval - 1 day);
      DATE_FORMAT(dp.InDateTime, '%d-%m-%Y %H:%i') dt,

**Comment by krasandr** [comment]
I confirm that yours last commit fixes example with IF with COUNT(*) to CASE above:

SET IsFertiliz = COALESCE((SELECT
IF (COUNT(*) = 0, 0, 1)
FROM ivf_stimulations s,
ivf_ins_fertilizations c,
ivf_ins_fertilizations_ivf_s_cells fc
WHERE s.IvfStimID = c.IvfStimID
AND fc.IvfInsFerID = c.IvfInsFerID
AND s.IvfArtID = IvfArtID), 0);
IF (IsFertiliz = 1)
THEN
RETURN 13;
END IF;

..
SET IsEmbrioDev = COALESCE((SELECT
IF (COUNT(*) > 0, 1, 0)
FROM v_ocells o
WHERE o.IvfArtID = IvfArtID
AND o.IvfOStatusID = 1), 0);

Now it is converted right to:

..
  SET IsFertiliz = COALESCE((SELECT
      CASE WHEN COUNT(*) = 0 THEN 0 ELSE  1 END.
    FROM ivf_stimulations s,
         ivf_ins_fertilizations c,
         ivf_ins_fertilizations_ivf_s_cells fc
    WHERE s.IvfStimID = c.IvfStimID
    AND fc.IvfInsFerID = c.IvfInsFerID
    AND s.IvfArtID = IvfArtID), 0);
  IF (IsFertiliz = 1)
    THEN
    RETURN 13;
  END IF;

..
  SET IsEmbrioDev = COALESCE((SELECT
      CASE WHEN COUNT(*) > 0 THEN 1 ELSE  0 END.
    FROM v_ocells o
    WHERE o.IvfArtID = IvfArtID
    AND o.IvfOStatusID = 1), 0);

  IF (IsEmbrioDev = 1)
    THEN
    RETURN 7;
  END IF;

I think that right now, the only one issue is that yours previous fix that removes comma from ",IN", must be applied to work in VIEWs definitions also (all of them are with ",IN"), so I am closing this bug for @variables. Thank you.

**Comment by krasandr** [comment]
It is my mistake, there is not many commas in VIEWs, the ",IN" is only when there is conversion of MySQL IF(a,b,c) form. For example:

 | v_deppatient_locations | CREATE ALGORITHM=UNDEFINED DEFINER=`root`@`%` SQL SECURITY DEFINER VIEW `v_deppatient_locations` AS select `dp`.`DepartmentPationID` AS `DepartmentPationID`,`c`.`CabinetName` AS `CabinetName`,`cb`.`CabinetBedName` AS `CabinetBedName` from (`department_patients` `dp` left join (((`patient_movements` `pm1` join `patient_movements` `pm2`) join `cabinet_beds` `cb`) join `cabinets` `c`) on(`pm1`.`DepartmentPationID` = `pm2`.`DepartmentPationID` and `pm1`.`CabinetID` = `c`.`CabinetID` and `pm1`.`CabinetBedID` = `cb`.`CabinetBedID` and `pm1`.`DepartmentPationID` = `dp`.`DepartmentPationID`)) where `dp`.`PatientStatusID` in (6,8,9) group by `dp`.`DepartmentPationID`,`pm1`.`DepartmentPationID`,`pm1`.`PatientMovementID` having `pm1`.`PatientMovementID` = max(`pm2`.`PatientMovementID`) or `pm1`.`PatientMovementID` is null | utf8                 | utf8_general_ci      |

And another example with same ",IN" in VIEW, again related only to IF(a,b,c) conversion:

| v_patient_movment_helper1 | CREATE ALGORITHM=UNDEFINED DEFINER=`root`@`%` SQL SECURITY DEFINER VIEW `v_patient_movment_helper1` AS select `mv`.`PatientMovementID` AS `PatientMovementID`,`mv`.`DepartmentPationID` AS `DepartmentPationID`,`mv`.`MovementTime` AS `MovementTime`,`mv`.`MovementTypeID` AS `MovementTypeID`,`mv`.`CabinetID` AS `CabinetID`,`mv`.`CabinetBedID` AS `CabinetBedID`,`mv`.`IsIndependent` AS `IsIndependent`,`mv`.`IsReserve` AS `IsReserve`,`mv`.`DepartmentID` AS `DepartmentID` from (`patient_movements` `mv` join `department_patients` `dp`) where `dp`.`DepartmentPationID` = `mv`.`DepartmentPationID` and `dp`.`PatientStatusID` in (6,8) and `mv`.`MovementTypeID` in (1,2) and `mv`.`PatientMovementID` = (select max(`pmv`.`PatientMovementID`) AS `MAX(``pmv``.``PatientMovementID``)` from `patient_movements` `pmv` where `pmv`.`DepartmentPationID` = `dp`.`DepartmentPationID` and `pmv`.`MovementTypeID` in (1,2)) | utf8                 | utf8_general_ci      |

After conversion, the two VIEWs are converted with ",IN (something)", with comma before IN.

**Comment by krasandr** [comment]
I comment in yours previous commit for ",IN" fix:
There is a problem with this commit, it don't work, for example in VIEW, if "in" is with small letter.
If I change it to 
$str =~ s/\s*,\s*in\s*\(/ in \(/gs;
it removes commas from ",in", so right solution will be non case-sensitive regex.

**Comment by krasandr** [comment]
After your last commit, all DATE_FORMAT, DATE_ADD, LAST_DAY are converted correctly, thank you.
There is only one, that is not converted, but it is really very complex, so I think, it is normal exception.
If you are interested, here is the only one exception:
| f_getlastedit | STRICT_TRANS_TABLES,NO_AUTO_CREATE_USER,NO_ENGINE_SUBSTITUTION | CREATE DEFINER=`root`@`%` FUNCTION `f_getlastedit`(pLastUpdateTime datetime, pUserID int(11)) RETURNS text CHARSET utf8
BEGIN

  IF (pLastUpdateTime IS NULL
    OR pUserID IS NULL) THEN
    RETURN '';
  END IF;

  SET @UserName := '';
  SET @DoctorName := '';

  SELECT
    u.UserName,
    F_GetDoctorName(u.DoctorID) INTO @UserName,
  @DoctorName
  FROM users u
  WHERE u.UserID = pUserID;

  RETURN CONCAT('Last redaction ', DATE_FORMAT(IFNULL(pLastUpdateTime, NOW()), '%d.%m.%Y %H:%i'), 'hour from ', @DoctorName, ' (', @UserName, ')');

END                      | utf8                 | utf8_general_ci      | latin1_swedish_ci  |

By the way, everything works O.K., except what I write before about case sensitive ",in" with small letters, in which comma is not removed, based on regex.

**Comment by darold** [comment]
Ok, I think that last few commits fixes the pending and new issues you've reported. Please give it a try and let me know. Thanks for all your reports, this helps a lot to improve the MySQL code translation.

**Comment by krasandr** [comment]
I confirm that everything is converted right after last commit, thank you very much.
Right now, I cannot find any problem with variables, views, functions, procedures and triggers, but tomorrow I will test some kind of automation of conversion with detailed debug, and also translation of type query (from input file) and some binary objects store in database (clob,blob) and I will report to you. Also, there is possibility that I miss some MySQL string function that is not available in PostgreSQL, but I will check again all.

**Comment by darold** [comment]
Great, I have never tested clob/blob export on MySQL so there will be some issues, also my fisrt test on data export shows that MySQL is very slow at exporting data, but I have not investigate more. Your feedbacks will be very appreciated.
